[4.0] Namespace - Change default controller

joomdonation opened the pull request:
Pull Request for Issue #15041.

### Summary of Changes
This PR changes default controller name from COM_NAME to a generic **controller** name as discussed in #15041. I also make some small clean up to the code which can be easily seen while review

### Testing Instructions
Code review. Ping @wilsonge and @mbabker for final decision

joomdonation committed:
[4.0] Namespace - Change default controller

joomdonation commented:
While looking at this, I also have some questions:

1. With the current implement, the data stored in namespace column must contain backslash Joomla\Component\Content\ . I think without backlash Joomla\Component\Content might be better

2. Do we really want to introduce a new way to pass controller name via controller variable in the request? At least for current usage, the widely use task=controller.method is good enough.

joomdonation committed:
Code style

laoneo commented on libraries/src/Cms/Controller/Admin.php:
```diff
@@ -3,8 +3,8 @@
  *
- * @copyright  Copyright (C) 2005 - 2017 Open Source Matters, Inc. All rights reserved.
- * @license    GNU General Public License version 2 or later; see LICENSE.txt
+ * @copyright   Copyright (C) 2005 - 2017 Open Source Matters, Inc. All rights reserved.
```
Why this changes, only 2 spaces should be there, otherwise you get CS errors.

joomdonation replied:
It was CS error before I made the change :(

laoneo replied:
It must be only two spaces, when you start to revert all of that, then we need to redo this changes again. I don't know why this gives errors on j4, but on 3.8 it is correct and that's how the flow should be staging -> 3.8 -> 4.0.

joomdonation replied:
Someone who setup code style rules for 4.0-dev branch has to check it then. Otherwise, every PR to 4.0-dev will get this codestyle error.

However, quick look at it again, maybe 3 space is expected change:

1. It make the copyright text align property with the text in @package and @subpackage tag

2. Look at this file https://github.com/joomla/joomla-cms/blob/staging/libraries/cms/application/administrator.php#L6 for example, it is also use 3 space for @copyright tag.

laoneo replied:
The package information is not required for the namespaced classes. The namespaced classes do have a different doc block than the none namespaced ones.

joomdonation replied:
OK. I changed the doc block of these classes to make it the same with other namespaced classed. CS looks OK now

joomdonation committed:
Revert "Code style"

This reverts commit 761fd8c3e239aa69bc461a4642aaa484a038fd6e.

joomdonation committed:
Remove @package and @subpackage  tags, fix CS

RonakParmar commented on libraries/src/Cms/Dispatcher/Dispatcher.php:
```diff
@@ -163,11 +163,16 @@ protected function getApplication()
+		$client = $client ? $client : ucfirst($this->app->getName());
 
-		$controllerName = $this->namespace . $client . 'Controller\\' . ucfirst($name);
+		$controllerClass = $this->namespace . $client . '\\Controller\\' . ucfirst($name);
```
In old code, '\\' was appended to only ucfirst($this->app->getName()) value, and not to the $client value.
But the current code is appending '\\' to both the values "ucfirst($this->app->getName())" and "$client".
Is it correct to append '\\' to the $client value?

joomdonation replied:
Yes. That's correct. $client if passed, will have Site or Administrator value, so append '\' is correct

RonakParmar commented:
I have tested this item :white_check_mark: successfully on b47213c2cfc1da8ce841ca693210cb61c71d55b9<br /><br />Done code review.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/15075">issues.joomla.org/tracker/joomla-cms/15075</a>.</sub>

joomdonation committed:
Component namespace should not contain backslash

joomdonation committed:
Change Controller to BaseController

Avoid having to write long code like use
Joomla\Cms\Controller\Controller as CmsController in default component
controller

joomdonation committed:
Pass option and name  config option to controller contructor

Avoid controller has to dynamic calculate it from class name

joomdonation committed:
CS

joomdonation committed:
Restore Controller class

joomdonation committed:
Redo change Controller to BaseController

joomdonation committed:
Update class name

joomdonation committed:
Remove this test for now. Guess it was written for none-namepace class

joomdonation commented:
So I made further change beside to just changes default controller name;

1. Change Controller to BaseController. It will allows us to write shorter code in default component controller

use Joomla\Cms\BaseController;
class Controller extends BaseController;

Instead of 

use Joomla\Cms\Controller as CmsController;
class Controller extends CmsController;

2. Passing option and name to controller config array. That save controller not having to detect dynamic these information from class name.

joomdonation commented:
Again, close because I could not resolve conflicts :(
